runtime/sqldelight-runtime/src/main/java/com/squareup/sqldelight/Transacter.kt

```diff
@@ -26,12 +27,12 @@ abstract class Transacter(private val helper: SqlDatabase) {
    * For internal use, performs [function] immediately if there is no active [Transaction] on this
    * thread, otherwise defers [function] to happen on transaction commit.
    */
-  protected fun notifyQueries(queries: List<Query<*>>) {
+  protected fun notifyQueries(queryList: QueryList) {
     val transaction = helper.getConnection().currentTransaction()
     if (transaction != null) {
-      transaction.queriesToUpdate.addAll(queries)
+      transaction.queriesToUpdate.addAll(queryList.queries)
     } else {
-      queries.forEach { it.notifyResultSetChanged() }
+      queryList.queries.forEach { it.notifyResultSetChanged() }
```

Comment:
This is a lot more safe than the locked version, but I think there's still a potential for a concurrent modification exception to occur if the of running a query causes someone to add/remove a query from this list.

Reply:
Or even just unlucky timing of someone subscribing while this is iterating...

Reply:
I'm trying to think of a way of going full on copy-on-write pattern but it's hard to see.

https://github.com/google/guava/blob/2cac83e70d77f0fa9b2352fe5ac994280fc3b028/guava/src/com/google/common/eventbus/SubscriberRegistry.java#L59-L62

https://github.com/Kotlin/kotlinx.coroutines/blob/379f210f1d6f6ee91d6198348bd16306c93153ce/core/kotlinx-coroutines-core/src/main/kotlin/kotlinx/coroutines/experimental/channels/ArrayBroadcastChannel.kt#L67

Reply:
i dont think concurrent mod is possible - theres no mutable list used anywhere. Every operation is currently creating a new list.

Reply:
Oh yeah. For some reason I thought it was a mutable list. Probably reviewing too late at night.